stage0/src/image_header.rs

```diff
@@ -2,155 +2,106 @@
-    Some(imagea)
-}
-
-// The careful observer will note that yes this is just the
-// start of an ARMv8m image with extra data shoved in the
-// vector table
-#[repr(C)]
-pub struct ImageHeader {
-    sp: u32,
-    pc: u32,
-    _vector_table: [u8; 24],
-    image_length: u32,
-    _image_type: u32,
-    header_offset: u32,
+    Some(img)
 }
 
-impl ImageHeader {
-    pub fn get_img_start(&self) -> u32 {
-        self as *const Self as u32
-    }
-
-    fn check_bounds(&self, address: u32) -> bool {
-        let start = self.get_img_start();
-
-        return address >= start && address < (start + self.image_length);
+impl Image {
+    fn get_img_start(&self) -> u32 {
+        self.0 as *const ImageVectors as u32
     }
 
-    fn get_table_start(&self) -> u32 {
-        self.get_img_start() + self.header_offset
+    fn get_header(&self) -> *const ImageHeader {
+        // SAFETY: This generated by the linker script which we trust
+        // Note that this is generated from _this_ image's linker script
+        // as opposed to the _image_ linker script but those two _must_
+        // be the same value!
+        let vector_size = unsafe { core::ptr::addr_of!(__vector_size) as u32 };
+        (self.get_img_start() + vector_size) as *const ImageHeader
     }
 
     /// Make sure all of the image flash is programmed
-    fn validate(&self) -> bool {
+    pub fn validate(&self) -> bool {
         let img_start = self.get_img_start();
 
-        // Start by making sure the region is actually programmed
+        // Start by making sure we can access the page where the vectors live
         let valid = lpc55_romapi::validate_programmed(img_start, 0x200);
 
         if !valid {
             return false;
         }
 
+        let header_ptr = self.get_header();
+
+        // Next validate the header location is programmed
+        let valid = lpc55_romapi::validate_programmed(header_ptr as u32, 0x200);
+
+        if !valid {
+            return false;
+        }
+
+        // SAFETY: We've validated the header location is programmed so this
+        // will not trigger a fault. This is generated from our build scripts
+        // which we trust.
+        let header = unsafe { &*header_ptr };
+
         // Next make sure the marked image length is programmed
         let valid = lpc55_romapi::validate_programmed(
             img_start,
-            (self.image_length + 0x1ff) & !(0x1ff),
+            (header.total_image_len + 0x1ff) & !(0x1ff),
         );
 
         if !valid {
             return false;
         }
 
+        // Does this look correct?
+        if header.magic != abi::HEADER_MAGIC {
+            return false;
+        }
+
         return true;
     }
 
-    fn get_cert_table(&self) -> *const CertHeader {
-        self.get_table_start() as *const CertHeader
+    pub fn get_vectors(&self) -> u32 {
+        self.get_img_start()
     }
 
-    pub fn get_pc(&self) -> u32 {
-        self.pc
+    pub unsafe fn get_pc(&self) -> u32 {
```

Comment:
So, in an uncharacteristic comment for me, I don't think this needs to be `unsafe`. Here's my argument.

- `Image`'s only field the reference to `ImageHeader`, is private to this module, and you are very careful to only build it with valid information (sourced from the build system and `validate`-d). This means anyone holding an `Image` is holding positive proof that the header can be accessed correctly, I believe.
- Because it's a reference (`&'static ImageHeader`) we're telling the Rust compiler that it's initialized and safe to dereference, so any operation on `self.0` should be ok. (We can do this because of your careful validation above.)

Whatcha think?